### DISCUSSION TITLE
Help with extending in JavaScript (xml)

### DISCUSSION THREAD

**Discussion by emrahamet** [pocketbase/pocketbase#3757] [Q&A] [answered]
While extending with JavaScript, I want to download and process an XML file, but I couldn't convert the XML into a usable form using JavaScript. I tried some modules (), but the result didn't change. What I want to do is download the XML file using $http.send, process it, and then save it to a database.

xml2js
xml-js
fast-xml-parser

**Comment by ganigeorgiev** [comment]
What do you mean by _"the result didn't change"_?
Could you provide a code sample and post the exact error that is returned?

**Reply by ganigeorgiev** [reply, marked as answer]
I haven't tested the above JS libraries but as long as they don't have dependency on some browser or node specific APIs, they should work.

I've only quickly checked the code of [`xml-js`](https://github.com/nashwaan/xml-js/tree/master) and at first glance I don't see anything platform specific.
To import it you must have a `node_modules` directory somewhere in your current parent or working directory with the module downloaded in it (eg. with `npm install`).
Alternatively, you can also try downloading their [dist file](https://github.com/nashwaan/xml-js/tree/master/dist) and load it directly but note that based on their [webpack.config](https://github.com/nashwaan/xml-js/blob/master/webpack.config.js#L13) those dist files are generated for a browser/window target, aka. if you go with this route then you'll have to define an empty global `window` object in your handler in order to work. Here is a quick example with manually loading a local [`xml-js.js` dist file](https://raw.githubusercontent.com/nashwaan/xml-js/master/dist/xml-js.js):

```js
// download one of the dist files into pb_hooks/xml-js.js and then require it...

this.window = {};
require("./pb_hooks/xml-js.js")

const res = $http.send(...)

// for all options check the package documentation
// https://www.npmjs.com/package/xml-js#options-for-converting-xml-%E2%86%92-js-object--json
const obj = this.window.xml2js(res.raw, { compact: true, nativeType: true });

console.log(JSON.stringify(obj))
```

(_I haven't tested the other solutions, nor whether the above library handles all cases_)

**Reply by ganigeorgiev** [reply]
_Side-note: Keep in mind that these libraries are huge and the PocketBase JSVM performance and resource consumption may not be that great when non-go wrapped code is being executed. If that ended up a bottleneck for your application, I recommend consider migration your code to [Go](https://pocketbase.io/docs/go-overview/) or process your data in a JS focused runtime (eg. node/deno/bun) and interact with Pocketbase as admin._

**Reply by emrahamet** [reply]
Couldn't achieve it with node_modules, but when I directly copy-pasted it, it worked. However, it's using too much CPU

Do you have any recommended Go module?

**Reply by ganigeorgiev** [reply]
The std [`encoding/xml`](https://pkg.go.dev/encoding/xml) usually should be enough for most cases, but it really depends what you are trying to do.

**Reply by emrahamet** [reply]
My initial idea was as follows: 
sending a request with cron every hour, 
fetching the content of the XML (4 dif xml's), 
comparing it with the database to find changes,
 and sending notifications via email. 
This way, I would have had a database refreshed every hour in my possession.

**Comment by emrahamet** [comment]
xmljs.pb.js

```js
// 1. Dowload Sax => https://github.com/isaacs/sax-js (zip or git clone)
// 2. Change 'var sax = require('sax');' to 'var sax = require('./sax');' in xml2js.js

routerAdd("GET", "/xmljs", (c) => {

    const res = $http.send({
        url: "https://example.com/products.xml",
        method: "GET",
        body: "", // eg. JSON.stringify({"test": 123})
        // headers: {"content-type": "application/json"},
        headers: {"content-type": "application/xml"},
        timeout: 120, // in seconds
    });

    if(res.statusCode != 200){

        return c.json(res.statusCode, {"code": res.statusCode, "message": "request failed"});
    }

    const convert = require(`${__hooks}/xmljs/index.js`);

    // https://www.npmjs.com/package/xml-js#options-for-converting-xml-%E2%86%92-js-object--json
    let options = {
        compact: true,
        nativeType: true,
        trim: true,
        ignoreAttributes: true,
        spaces: 0,
        // cdataKey: '_',
        ignoreDeclaration: true,
        // cdataFn: function(value, parentElement){},
        // textFn: function(value, parentElement){},
    };

    // let obj = convert.xml2js(res.raw, options);
    // or
    let json = convert.xml2json(res.raw, options);

    c.json(200, json)
});
```
